packages/go_router/lib/src/misc/stacked_navigation_scaffold.dart

```diff
@@ -0,0 +1,207 @@
+import 'package:flutter/widgets.dart';
+
+import '../state.dart';
+
+/// Transition builder callback used by [IndexStackShell].
+///
+/// The builder is expected to return a transition powered by the provided
+/// `animation` and wrapping the provided `child`.
+///
+/// The `animation` provided to the builder always runs forward from 0.0 to 1.0.
+typedef StackedNavigationTransitionBuilder = Widget Function(
+  BuildContext context,
+  Animation<double> animation,
+  Widget child,
+);
+
+/// Builder for the scaffold of a [StackedNavigationScaffold]
+typedef StackedNavigationScaffoldBuilder = Widget Function(
+    BuildContext context,
+    int currentIndex,
+    List<StackedNavigationItemState> itemsState,
+    Widget scaffoldBody);
+
+/// Representation of a item in the stack of a [StackedNavigationScaffold]
+class StackedNavigationItem {
+  /// Constructs an [StackedNavigationItem].
+  StackedNavigationItem(
+      {required this.initialLocation, required this.navigatorKey});
+
+  /// The initial location/path
+  final String initialLocation;
+
+  /// Optional navigatorKey
+  final GlobalKey<NavigatorState> navigatorKey;
+}
+
+/// Represents the current state of a [StackedNavigationItem] in a
+/// [StackedNavigationScaffold]
+class StackedNavigationItemState {
+  /// Constructs an [StackedNavigationItemState].
+  StackedNavigationItemState(this.item);
+
+  /// The [StackedNavigationItem] this state is representing.
+  final StackedNavigationItem item;
+
+  /// The current [GoRouterState] for the navigator of this item.
+  GoRouterState? currentRouterState;
+
+  /// The [Navigator] for this item.
+  Navigator? navigator;
+
+  /// Gets the current location from the [currentRouterState] or falls back to
+  /// the initial location of the associated [item].
+  String get currentLocation => currentRouterState != null
+      ? currentRouterState!.location
+      : item.initialLocation;
+}
+
+/// Widget that maintains a stateful stack of [Navigator]s, using an
+/// [IndexStack].
+class StackedNavigationScaffold extends StatefulWidget {
+  /// Constructs an [IndexStackShell].
+  const StackedNavigationScaffold({
+    required this.currentNavigator,
+    required this.currentRouterState,
+    required this.stackItems,
+    this.scaffoldBuilder,
+    this.transitionBuilder,
+    this.transitionDuration = defaultTransitionDuration,
+    super.key,
+  });
+
+  /// The default transition duration
+  static const Duration defaultTransitionDuration = Duration(milliseconds: 400);
+
+  /// The navigator for the currently active tab
+  final Navigator currentNavigator;
+
+  /// The current router state
+  final GoRouterState currentRouterState;
+
+  /// The tabs
+  final List<StackedNavigationItem> stackItems;
+
+  /// The scaffold builder
+  final StackedNavigationScaffoldBuilder? scaffoldBuilder;
+
+  /// An optional transition builder for stack transitions
+  final StackedNavigationTransitionBuilder? transitionBuilder;
```

Comment:
I think it's a good idea to make this optional as the default when switching tabs on iOS is no transition at all.

Reply:
Yeah, you're right. I guess I thought it would be ok since it isn't required and set to a default value, but I agree it's cleaner to just have it optional and handle the default internally. Thanks!